attalos/dataset/vg_prep.py

```diff
@@ -0,0 +1,176 @@
+            zipref.extractall(self.data_dir)
+
+        if extract_all_images:
+            zipref = zipfile.ZipFile(self.images_filename,'r')
+            zipref.extractall(self.data_dir)
+
+    def load_metadata(self):
+        """
+        Load the VGG dataset to allow for efficient iteration
+        Returns:
+
+        """
+        if self.split == SplitType.TRAIN:
+            split_name = 'train'
+        elif self.split == SplitType.VAL:
+            split_name = 'val'
+        else:
+            raise NotImplementedError('Split type not yet implemented')
+
+        self.item_info = json.loads(open(self.metadata_filename[:-4], 'r').read().decode('ascii'))
+        # self.item_info=lapi.GetAllImageData(dataDir=self.data_dir)
+        
+
+    def get_key(self, key):
+        """
+        Return metadata about key
+        Args:
+            key: ID who's metadata we'd like to extract
+
+        Returns:
+            RecordMetadata: Returns ParserMetadata object containing metadata about item
+        """
+        
+        # item = self.item_info[key]
+        # return RecordMetadata(id=key, images_name=item['fname'], tags=item['tags'], captions=item['captions'])
+
+        return self.item_info[key-1]
+
+    def extract_image_by_key(self, key):
+        """
+        Return an image based on the input key
+        Args:
+            key: ID of the file we'd like to extract
+
+        Returns:
+            Image Blob: Bytes of the image associated with the input ID
+        """
+        imname = self.item_info[key].url.split('/')[-1]
+
+        key_info = self.get_key(key)
+
+        if self.images_file_handle is None:
+            self.images_file_handle = zipfile.ZipFile(self.images_filename)
+
+        zipfile.ZipFile.namelist(self.images_file_handle)
+        print('image name {}'.format(self.images_filename))
+        train_captions = self.images_file_handle.open('%s'%imname)
+
+        return train_captions.read()
+
+    def extract_image_to_location(self, key, desired_file_path):
+        """
+        Write image based on the input key to the desired location
+        Args:
+            key: ID of the file we'd like to extract
+            desired_file_path: Output filename that we should write the file to
+
+        Returns:
+        """
+
+        fOut = open(desired_file_path, 'wb')
+        fOut.write(self.extract_image_by_key(key))
+        fOut.close()
+
+    def get_image_by_key(self, key):
+        ''' 
+        Gets the image (assuming it's been extracted).
+        '''
+        
+        imdata = self.get_key(key)
+        imurl = imdata['url'].split('/')[-1]
+        return self.data_dir+imurl
+
+    def __iter__(self):
+        """
+        Iterator over the dataset.
+        Returns:
+            RecordMetadata: Information about the next key
+        """
+        for key in sorted(self.list_keys()):
+            yield self.get_key(key)
+
+        raise StopIteration()
+
+    def list_keys(self):
```

Comment:
I'm getting an error when I try to call this function:
```
>>> vg = vg_prep.VGDatasetPrep('/local_data/visual_genome/')                                                                                                 
>>> keys = vg.list_keys()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "attalos/dataset/vg_prep.py", line 176, in list_keys
    return self.item_info.keys()
AttributeError: 'list' object has no attribute 'keys'
```
I'm not totally sure but it looks like maybe it should be something like:
```
return [ item['id'] for item in self.item_info]
```
I think this would then clarify the question about the "-1" above